src/plugins/graph/models/data-configuration-model.ts

```diff
@@ -587,106 +613,141 @@ export const DataConfigurationModel = types
+        });
+      }
+    },
     setDataset(dataset: IDataSet | undefined, metadata: ISharedCaseMetadata | undefined) {
-      self.actionHandlerDisposer?.();
-      self.actionHandlerDisposer = undefined;
       self.dataset = dataset;
       self.metadata = metadata;
-      self.filteredCases = [];
-      if (dataset) {
+      this.handleDataSetChange();
+    },
+    handleDataSetChange() {
+      self.actionHandlerDisposer?.();
+      self.actionHandlerDisposer = undefined;
+      if (self.dataset) {
         self.actionHandlerDisposer = onAnyAction(self.dataset, self.handleAction);
-        self.filteredCases[0] = new FilteredCases({
-          source: dataset, filter: self.filterCase,
-          onSetCaseValues: self.handleSetCaseValues
-        });
-        // make sure there are enough filteredCases to hold all the y attributes
-        while (self.filteredCases.length < self._yAttributeDescriptions.length) {
-          this._addNewFilteredCases();
-        }
-        // A y2 attribute is optional, so only add a new filteredCases if there is one.
-        if (self.hasY2Attribute) {
-          this._addNewFilteredCases();
-        }
       }
+      this.syncFilteredCasesCount(true);
       self.invalidateQuantileScale();
     },
     setPrimaryRole(role: GraphAttrRole) {
       if (role === 'x' || role === 'y') {
         self.primaryRole = role;
       }
     },
+    /**
+     * Remove all attributes currently in our lists.
+     */
     clearAttributes() {
-      self._attributeDescriptions.clear();
-      self._yAttributeDescriptions.clear();
-    },
-    setAttribute(role: GraphAttrRole, desc?: IAttributeDescriptionSnapshot) {
-
-      // For 'x' and 'y' roles, if the given attribute is already present on the other axis, then
-      // move whatever attribute is assigned to the given role to that axis.
-      if (['x', 'y'].includes(role)) {
-        const otherRole = role === 'x' ? 'y' : 'x',
-          otherDesc = self.attributeDescriptionForRole(otherRole);
-        if (otherDesc?.attributeID === desc?.attributeID) {
-          const currentDesc = self.attributeDescriptionForRole(role) ?? {attributeID: '', type: 'categorical'};
-          self._setAttributeDescription(otherRole, currentDesc);
-        }
+      // Clear the attributes one by one so that reactions can happen
+      while (self._yAttributeDescriptions.length) {
+        this.removeYAttributeWithID(self._yAttributeDescriptions[0].attributeID);
       }
+      for (const [role] of self._attributeDescriptions) {
+        this.removeAttributeFromRole(role as GraphAttrRole);
+      }
+    },
+    /**
+     * Unset any attribute currently in the given role.
+     * Not used for Y attributes, which have their own methods.
+     * @param role GraphAttrRole, not inluding 'y'
+     */
+    removeAttributeFromRole(role: GraphAttrRole) {
+      self._setAttributeDescription(role);
+    },
+    setRoleToAttributeDesc(role: GraphAttrRole, desc?: IAttributeDescriptionSnapshot) {
```

Comment:
My inclination would be to call this `setAttributeForRole` -- attribute descriptions are an implementation detail that need not be explicit in the action names.